pkg/api-server/resource_endpoints.go

```diff
@@ -532,17 +532,45 @@ func (r *resourceEndpoints) validateLabels(rm model.ResourceMeta) validators.Val
 		}
 	}
 
-	for _, k := range maps.SortedKeys(rm.GetLabels()) {
+	if r.mode != config_core.Global {
+		zoneTag, ok := resource.GetMeta().GetLabels()[mesh_proto.ZoneTag]
+		if ok && zoneTag != r.zoneName {
+			err.AddViolationAt(validators.Root().Key(mesh_proto.ZoneTag), fmt.Sprintf("%s label should have %s value", mesh_proto.ZoneTag, r.zoneName))
+		}
+	}
+
+	if r.descriptor.IsPluginOriginated && r.descriptor.IsPolicy {
+		r.validatePolicyRole(resource)
+	}
+
+	for _, k := range maps.SortedKeys(resource.GetMeta().GetLabels()) {
 		for _, msg := range validation.IsQualifiedName(k) {
 			err.AddViolationAt(validators.Root().Key(k), msg)
 		}
-		for _, msg := range validation.IsValidLabelValue(rm.GetLabels()[k]) {
+		for _, msg := range validation.IsValidLabelValue(resource.GetMeta().GetLabels()[k]) {
 			err.AddViolationAt(validators.Root().Key(k), msg)
 		}
 	}
 	return err
 }
 
+func (r *resourceEndpoints) validatePolicyRole(resource rest.Resource) validators.ValidationError {
+	var err validators.ValidationError
+	policyObject, ok := resource.GetSpec().(core_model.Policy)
+	if !ok {
+		return err
```

Comment:
you're returning an empty error here, I think it's easier to not handle cast because you're checking the descriptor

Reply:
changed